### PR TITLE
Converted shared settings script into a convention plugin + Added `foojay` toolchain plugin

### PR DESCRIPTION
**Pull request by jasonsparc**
This fixes issue #59 while making sure that the `foojay` toolchain plugin is configured in only one place. To accomplish that, the `shared.settings.gradle.kts` file was converted into a [precompiled script](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins) convention plugin.

### CODE REVIEW

**Comment by jasonsparc** [comment]
Feel free to customize the plugin IDs, filenames, filesystem location, etc.

File: gradle/shared-settings/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+plugins {
+    `kotlin-dsl`
+}
+
+dependencies {
+    implementation("org.gradle.toolchains:foojay-resolver:0.5.0") // https://github.com/gradle/foojay-toolchains
```
Comment (gmazzo) [inline gradle/shared-settings/build.gradle.kts:6]: All dependencies should be declared in the TOML file, to allow Dependabot to pick them up for automatic updates.

**Review by gmazzo — COMMENTED** [review]
Thanks for the PR! As general feedback, I'd like to keep the logic as simple as possible. Moving the script file as a dedicated included build will be the right choice, in case its content requires to (like adding plugins, or have more complex stuff).

I this is about just refining a couple of repositories in a single place, I think it's just simpler to keep in that file.

Regarding the plugin, I'd just put it in the TOML file and apply it in the `plugins` module, which will download the JDK the first time it's installed.

File: gradle/shared-settings/build.gradle.kts (hunk 1 above)
Comment (jasonsparc) [reply gradle/shared-settings/build.gradle.kts]: Alright. I pushed another commit.

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -1,7 +1,9 @@
 [versions]
 kotlin = "1.8.22"
+foojay = "0.5.0"
 
 [libraries]
+foojay = { module = "org.gradle.toolchains:foojay-resolver", version.ref = "foojay" }
```
Comment (gmazzo) [inline gradle/libs.versions.toml:6]:
```suggestion
foojay = { module = "org.gradle.toolchains:foojay-resolver", version.ref = "0.5.0" }
```
This is also opnionated, but I personally like to use `[versions]` when there is a good reason for it.

Comment (jasonsparc) [reply gradle/libs.versions.toml:6]: If I move the `foojay-jvmToolchain` entry into the `[plugins]` section, then I would have to apply it via,

```
plugins {
    alias(libs.plugins.foojay.jvmToolchain)
}
```

However, `libs` doesn't seem to be available in `settings.gradle[.kts]` files, including in `settings` plugins. Investigating further lead me to this issue, https://github.com/gradle/gradle/issues/24876 – and according to that, the Gradle team has no plans to support `libs` ever being accessible in the `settings` context. Quote:

> This is the intended behavior, and there’s no plan to change it. This issue will be closed.
>
> Version catalogs can't be used before they have been evaluated, which happens during settings evaluation, which is AFTER evaluating ***settings plugins***. So basically a chicken and egg problem; nothing we can do about it.

(Emphasis mine.)

Comment (jasonsparc) [reply gradle/libs.versions.toml:6]: Although, in my opinion, the Gradle team could've at least supported the `libs` variable in precompiled script plugins of `settings` plugins. Actually, there's even another issue for that: “[Make version catalogs accessible from precompiled script plugins · Issue #15383 · gradle/gradle](https://github.com/gradle/gradle/issues/15383)” – this one is still open and not yet implemented.

Comment (gmazzo) [reply gradle/libs.versions.toml]: You are right 🤦. You can't, it's an egg-chicken problem. Still, I don't feel comfortable doing all this refactoring just because of this. Thanks again for dedicating time, but giving the complexity the value added is just not worth it to me. If JDK 8 is too old that people simply don't install it anymore, I'd rather consider bumping it to JDK11/17.

**Comment by jasonsparc** [comment]
Also some summary:

- If we use a shared script plugin as before, we can't utilize the entry in the TOML (as `libs` isn't available in any `settings` context).
  - As described in my [comment from the issue](https://github.com/gmazzo/gradle-buildconfig-plugin/issues/59#issuecomment-1601020616).
  - Relevant Gradle issue: https://github.com/gradle/gradle/issues/24876
- I couldn't move the `foojay-jvmToolchains` entry into the `[plugins]` section of the TOML, due to the `libs` variable being unavailable in precompiled script plugins.
  - As described in my [reply for your suggestion](https://github.com/gmazzo/gradle-buildconfig-plugin/pull/60#discussion_r1237097173)
  - Relevant Gradle issue: https://github.com/gradle/gradle/issues/15383

File: gradle/libs.versions.toml (hunk 2 above)
Comment (jasonsparc) [reply gradle/libs.versions.toml]: Note also that, even if you bump it to JDK 11 or 17, and say I have JDK 19, the download logic of Gradle still needs to kick in, since Gradle wants exact versions of what you currently have installed.

Comment (jasonsparc) [reply gradle/libs.versions.toml]: Furthermore, you shouldn't bump the JVM bytecode output of your custom gradle plugin, since the output classes should stay simply in Java 8. I don't have a ticket to give to you from Jetbrains but Android Studio currently thinks that all Gradle scripts are Java 8. Last time I tried to bump the version of my scripts I got some funny errors about classpaths and Kotlin complaining that it couldn't inline a function whose bytecode is higher than the current bytecode (since it thinks the user's scripts are all Java 8 bytecodes).

You could however, downgrade Gradle to `7.X` again, and all this complications would disappear.

Comment (jasonsparc) [reply gradle/libs.versions.toml]: Actually, in my current project, I'm stuck with Gradle `8.0.2`, due to this issue: https://github.com/gradle/gradle/issues/25164

I've faced nightmares like that in the past, which is why I personally don't update Gradle that much often, especially in production.

**Comment by jasonsparc** [comment]
Also, do you really not want to use an included build? As far as I know, the Gradle team doesn't seem to be putting that much effort in supporting [script plugins](https://docs.gradle.org/current/userguide/plugins.html#sec:script_plugins) – hence, the various issues we're facing now.

Many have migrated to convention plugins and custom build logic, due to it being regarded as "best" practice, even though in my opinion, a custom build logic (including [precompiled script plugins](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins)) is far more complicated and slower than a simple [script plugin](https://docs.gradle.org/current/userguide/plugins.html#sec:script_plugins). But then, that's the trend now (and I personally hate it actually). In addition to that, IDEs have poor support for script plugins.

P.S. Personally, if I have the choice, I would've opted for a simpler solution too – if only there aren't gotchas; and those gotchas wouldn't exist if it weren't for those so called best practices being regarded too highly even when not thoroughly researched nor scientifically proven 😛

**Comment by gmazzo** [comment]
Sorry for the delay, I took some time to analyze this from a distance.

So, basically, the goal of this PR is to have 0 friction for a first contributor that for any reason, does not have a JDK8 installed. 

From my point of view, I'd not expect a new contribution to panic when opening this project it fails to sync/build, as if they are reaching this project they will be mature enough to handle a JDK installation.

Anyway, let's say we still want to cover this edge case, given the restrictions we have with Gradle and the fact I'd also want to honor the pattern of declaring all dependencies on a TOML file, the only feasible path to follow is to add a second layer of included build as you did on this PR.

That for me is a no-go, not because it's not the correct approach given the restrictions, but because the increased complexity of trying to solve a problem that, given the statement of the begging, is not a real problem to me.

Said of that, I noted your willingness to contribute to this project and I'm thank full to you for that. So let's do this, I tried that just adding the plugin in `plugin/settings.gradle.kts` will do the work. I'll be fine to relax the TOML restriction for this one, in favor of a simpler build approach. Can you update this PR to do just that?

Basically revert all the changes and add 
```patch
Subject: [PATCH] Added default JVM toolchain provision
---
Index: plugin/settings.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugin/settings.gradle.kts b/plugin/settings.gradle.kts
--- a/plugin/settings.gradle.kts	(revision c30b2e9ba14c3e6f9c04ef35f495915358eff953)
+++ b/plugin/settings.gradle.kts	(date 1687507567797)
@@ -1,3 +1,7 @@
+plugins {
+    id("org.gradle.toolchains.foojay-resolver-convention") version "0.5.0"
+}
+
 apply(from = "../gradle/shared.settings.gradle.kts")
 
 dependencyResolutionManagement {
```

**Comment by jasonsparc** [comment]
> So, basically, the goal of this PR is to have 0 friction for a first contributor that for any reason, does not have a JDK8 installed.

Actually, the goal of this PR is to conform to the new Gradle requirement of *always* specifying a toolchain repository *whenever you use* Gradle's JVM toolchain support. This is documented by Gradle's migration guide: “[Upgrading your build from Gradle 7.x to 8.0 | Gradle User Manual](https://docs.gradle.org/8.1.1/userguide/upgrading_version_7.html#using_automatic_toolchain_downloading_without_having_a_repository_configured)”

In other words, upgrading from Gradle `7.x` to `8.0` was a breaking change, and this PR (along with issue #59) tries to solve that oversight by attempting to conform to that new *Gradle version __migration__ requirement*.

> Can you update this PR to do just that?
>
> Basically revert all the changes and add […]

Now, regarding your proposed solution, that would work, so long as the demo project also uses the same JVM toolchain target, (or no JVM toolchain was set up for it at all). However, right now, your Android demo project is set up to use its own JVM toolchain, i.e., JDK 11, as in this [line](https://github.com/gmazzo/gradle-buildconfig-plugin/blob/v4.1.0/demo-project/kts-android/build.gradle.kts#L9), which is different from the JVM toolchain used by the plugin's build script (declared by this [line](https://github.com/gmazzo/gradle-buildconfig-plugin/blob/v4.1.0/plugin/build.gradle.kts#L15)). Don't worry though, since that additional issue can be solved easily as follows:

A quick survey of your Android demo project tells me that you currently *do not need* a separate JVM toolchain for it, and that you only want it to output “JVM 11”-compatible Java classes. Not only that, as far as I can tell, your Android demo project isn't even using "JDK 11"-specific features, and thus, it might as well use JDK 8 (which seems to be AGP's default anyway).

Given the reasoning above, I've evaluated 3 options (two of which are quite simple) for solving the above-mentioned issue:

**Option 1:** In addition to your proposal, remove the JVM toolchain declaration from the Android demo project, since AGP, by default, uses JDK 8 anyway – thus making it the same as your plugin's JVM toolchain declaration.

<details>
<summary>Patch (in addition to your proposed patch)</summary>

```diff
Index: demo-project/kts-android/build.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/demo-project/kts-android/build.gradle.kts b/demo-project/kts-android/build.gradle.kts
--- a/demo-project/kts-android/build.gradle.kts(revision 987354a97a70dff397da9e1ab11f2f86d0f3c6e7)
+++ b/demo-project/kts-android/build.gradle.kts(date 1687517919084)
@@ -6,8 +6,6 @@
     id("com.github.gmazzo.buildconfig")
 }
 
-java.toolchain.languageVersion.set(JavaLanguageVersion.of(11))
-
 android {
     namespace = "com.github.gmazzo.buildconfig.demos.android"
     compileSdkVersion = "android-30"
```

</details>

**Option 2:** In addition to your proposal, let the Android demo project use the same JVM toolchain as what the plugin's build had declared. That way, the plugin's build script would download JDK 8 first, and then the demo project's build can pick that up instead (from Gradle's `jdk` cache folder).

<details>
<summary>Patch (in addition to your proposed patch)</summary>

```diff
Index: demo-project/kts-android/build.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/demo-project/kts-android/build.gradle.kts b/demo-project/kts-android/build.gradle.kts
--- a/demo-project/kts-android/build.gradle.kts(revision 987354a97a70dff397da9e1ab11f2f86d0f3c6e7)
+++ b/demo-project/kts-android/build.gradle.kts(date 1687518678659)
@@ -6,7 +6,7 @@
     id("com.github.gmazzo.buildconfig")
 }
 
-java.toolchain.languageVersion.set(JavaLanguageVersion.of(11))
+java.toolchain.languageVersion.set(JavaLanguageVersion.of(8))
 
 android {
     namespace = "com.github.gmazzo.buildconfig.demos.android"
```

</details>

**Option 3:** Instead of using Gradle's JVM toolchain support, just configure the `JavaCompile` and `KotlinCompile` tasks to output a consistent Java bytecode output. And remove the JVM toolchain declarations.

I have not evaluated option 3 well (as I'm still about to do that later), but it might actually be more correct, and probably simple enough that you might like it too.

The reason why I've thought of option 3 is because, technically, the `javac` of higher JDKs should be able to produce outputs for lower Java class bytecodes. Therefore, there should be no need to download (or even install) two separate JDKs just to build your project (e.g., JDK 8 and JDK 11, in addition to the current user's installed JDK).

---

Further background regarding **option 3**:

In my opinion, Gradle's JVM toolchain support is meant to simulate a certain JVM environment, by making sure that the tools the build uses (such as `javac`) are never newer than the target environment. That is, it is meant to improve reproducibility of the build; quote from the [Gradle docs](https://docs.gradle.org/8.1.1/userguide/toolchains.html#:~:text=By%20default%2C%20Gradle%20uses%20the%20same%20Java%20toolchain%20for%20running%20Gradle%20itself%20and%20building%20JVM%20projects.%20However%2C%20this%20may%20only%20sometimes%20be%20desirable.):

> By default, Gradle uses the same Java toolchain for running Gradle itself and building JVM projects. However, this may only sometimes be desirable. […]
>
> In order to ***improve reproducibility*** of the builds and make build requirements clearer, Gradle allows configuring toolchains on both project and task levels.

(Emphasis mine.)

Hence, using Gradle's JVM toolchain support will actually download entire JDKs: one for Java 8 (as in this [line](https://github.com/gmazzo/gradle-buildconfig-plugin/blob/v4.1.0/plugin/build.gradle.kts#L15)), and the other for Java 11 (as in this [line](https://github.com/gmazzo/gradle-buildconfig-plugin/blob/v4.1.0/demo-project/kts-android/build.gradle.kts#L9)).

However, as far as I can tell, your project currently doesn't really need Gradle's JVM toolchain support. A quick survey of the build tells me that all you really want is to make sure that the output bytecodes are, Java 8, for the plugin artifact, and Java 11, for the Android demo project.

Now, all JDKs are backwards compatible. The `javac` of JDK 19 for example, can easily output, not only Java 19 bytecode, but also, Java 17, Java 11 and Java 8 bytecodes. Therefore, using an entire JVM toolchain, just to compile classes for Java 8 compatibility is actually overkill, don't you think?

Gradle's toolchain support downloads an entire JDK, only for the build to use its `javac` to compile stuffs that the user's current JDK's `javac` can already do, and the rest is wasted bandwidth and disk space. It's like using a sledgehammer just to nail down a single tiny nail: an entirely separate JDK, just to use its `javac`!

**Option 3** thus proposes to abandon the JVM toolchain declarations, in favor of simply configuring the various `JavaCompile` and `KotlinCompile` tasks for a consistent Java bytecode outputs. I haven't done that yet, but I might try that if you want.

**Comment by jasonsparc** [comment]
Oops! I forgot that you were manually setting up the `android.compileOptions {}` block as well, so here's **option 1** with an updated patch:

```diff
Index: demo-project/kts-android/build.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/demo-project/kts-android/build.gradle.kts b/demo-project/kts-android/build.gradle.kts
--- a/demo-project/kts-android/build.gradle.kts	(revision 987354a97a70dff397da9e1ab11f2f86d0f3c6e7)
+++ b/demo-project/kts-android/build.gradle.kts	(date 1687528346807)
@@ -6,18 +6,13 @@
     id("com.github.gmazzo.buildconfig")
 }
 
-java.toolchain.languageVersion.set(JavaLanguageVersion.of(11))
-
 android {
     namespace = "com.github.gmazzo.buildconfig.demos.android"
     compileSdkVersion = "android-30"
 
     buildFeatures.buildConfig = true
 
-    compileOptions {
-        targetCompatibility(java.targetCompatibility)
-        sourceCompatibility(java.sourceCompatibility)
-    }
+    kotlinOptions.jvmTarget = "${compileOptions.targetCompatibility}"
 }
 
 dependencies {
```

Note that, even without the above modifications, `gradlew :demo-project:kts-android:assemble` works just fine, since both `targetCompatibility` and `sourceCompatibility` (of the `compileOptions {}` block) were set to `java.targetCompatibility` and `java.sourceCompatibility` respectively, which is equivalent to setting them to whatever is the Java version running the Gradle wrapper.

Now, the `targetCompatibility` and `sourceCompatibility` of the `android.compileOptions` block are both `JavaVersion.VERSION_1_8` by default, but `kotlinOptions.jvmTarget` is not (at least in recent versions of the Kotlin gradle plugin), and is instead set to `null`, which is equivalent to `kotlinOptions.jvmTarget = "${java.targetCompatibility}"` – previously, the docs for the `kotlinOptions.jvmTarget` says that it's set to the string `"1.8"` by default (but now, it says it's `null` by default). Hence, the manual setup above is necessary to keep `compileOptions` and `kotlinOptions` in sync.

Even so, **option 1** still remains as the simplest solution (without a lot of gotchas), don't you think?

**Comment by gmazzo** [comment]
> you still need to keep the one from the demo project in-sync

That yes can be done with a `[version]` entry on the TOML file, right? But yes, both will have to be bump to `11` as it's the minimum now on Android (that's why the demo project is at 11)


To be honest, I don't fully get if `targetCompatibility` and `sourceCompatibility` have to be replaced with `toolchain` or they serve different use cases. But what I don't like is to manually configure them for java, kotlin and android. The three of them seems to support wiring from `toolchain` and that's why I went that way.

Anyway, undoing the `toolchain` path will not fix anything the missing JDK thing (that a compatible one will automatically downloaded) but it will just make it less strict, as you pointed out, any JDK11+ will work to build the project. 
I'm wondering if there is any setting on `toolchain` to allow to pick any `JDK11+` compatible instead of just `JDK11` 🤔

**Comment by jasonsparc** [comment]
Actually, speaking for myself, as an Android developer, I can confirm that the minimum JDK *was* indeed 11, but now it's 17. See, https://developer.android.com/build/releases/gradle-plugin#8-0-0

However, that only applies for the JDK installation that you're using to build the project. The default output artifact is still Java 8, and in fact, you can even target Java 6 and still release an app for newer Android devices – both the device and AGP won't care.

In fact, try setting your JVM toolchain declaration to 17, then set your `JAVA_HOME` to anything less than JDK 17, then finally, build your Android demo project: AGP will cause the build to fail, complaining something like:

```
FAILURE: Build failed with an exception.

* What went wrong:
An exception occurred applying plugin request [id: 'com.android.application', version: '8.0.2']
> Failed to apply plugin 'com.android.internal.application'.
   > Android Gradle plugin requires Java 17 to run. You are currently using Java 11.
      Your current JDK is located in <some JDK 11 path>
      You can try some of the following options:
       - changing the IDE settings.
       - changing the JAVA_HOME environment variable.
       - changing `org.gradle.java.home` in `gradle.properties`.
```

In the above, I set the JVM toolchain to 17, while I set `JAVA_HOME` to a JDK 11 installation, and then I ran `gradlew help`. It failed. The JVM toolchain declaration couldn't do anything, since AGP already failed the build before the toolchain directive can even be read.

---

Basically, what we need to do is to simply write in the README that, in order to build the demos, or at least the Android demo, JDK 11 or 17 is required. However, given that you've bumped AGP to `8.0.2` (as in this [line](https://github.com/gmazzo/gradle-buildconfig-plugin/blob/87d55988872eb2536bb610575a7f1b324a06e6d9/gradle/libs.versions.toml#L12)), the requirement is now JDK 17.

As for the Android demo, it can simply output whatever is set as the default by AGP. You don't really have to configure a separate toolchain for it. Setting a toolchain globally in the build automatically sets `java.targetCompatibility` and `java.sourceCompatibility` for convenience, but all AGP really cares about is its `android.compileOptions` and `android.kotlinOptions`, and not the toolchain. That's why you had to set `android.compileOptions` manually even when you've already set the Java version in the toolchain.

I believe, AGP doesn't even use `javac` for compiling classes, as it compiles them directly to Android dex bytecodes – now there was a time when that wasn't always the case, but those times are pretty much gone now. In other words, it really has no use for the toolchain declaration.

---

> Anyway, undoing the toolchain path will not fix anything the missing JDK thing (that a compatible one will automatically downloaded) but it will just make it less strict, as you pointed out, any JDK11+ will work to build the project. I'm wondering if there is any setting on toolchain to allow to pick any JDK11+ compatible instead of just JDK11 🤔

I believe, there is none (at the moment).

In fact, reading the Gradle docs tells me that, the purpose of the JVM toolchain is to strictly enforce a specific JDK tool set or strict JVM runtime environment (and we know that Android doesn't run in JVM). The only thing that's flexible for now is, which vendor and implementation (e.g., OpenJ9 or vendor-specific) you would like to *strictly enforce* for your target JVM environment. Future versions of Gradle might allow further customizations of the download logic, but for now, according to the [docs](https://docs.gradle.org/8.1.1/userguide/toolchain_plugins.html):

> Toolchain resolver plugins provide logic to map a toolchain request to a download response. At the moment the download response only contains a download URL, but may be extended in the future.
>
> > [!] For the download URL only secure protocols like `https` are accepted. This is required to make sure ***no one can tamper*** with the download in flight.

(Emphasis mine.)

Hence, from what I can tell, and with the above, the goal of Gradle's JVM toolchain support is to make the JVM environment more strict (and not more lax). The automatic `java.targetCompatibility` and `java.sourceCompatibility` being set is simply a convenient bonus.

Sadly, a lot of people are probably misunderstanding what Gradle's JVM toolchain support is really for.

**Comment by jasonsparc** [comment]
> However, that only applies for the JDK installation that you're using to build the project. The default output artifact is still Java 8, and in fact, you can even target Java 6 and still release an app for newer Android devices – both the device and AGP won't care.

Also, for clarity, what I meant was that, even if AGP now requires JDK 17+, it doesn't mean that you must also set your `android.compileOptions.targetCompatibility` and `android.compileOptions.sourceCompatibility` to be at least 17. In fact, you'd want to do the opposite instead: you should set them to be always lower than AGP's new JDK requirement – that's the reason why AGP set them to Java 8 by default, as that's the lowest reasonable target.

In fact, setting them too high is probably a bad idea, as Android's dexer doesn't seem to support all Java language features, especially those of higher Java versions (since Android is not a JVM anyway).

Generally, you'd want them to be set to something like less than or equal to Java 11. That is, it's unrelated to AGP's JDK requirement, since the JDK is only used to run AGP and not for compiling the Android sources.

**Comment by jasonsparc** [comment]
I think you could use JDK 17 as the JVM toolchain of the `plugin` project, and (at the same time), have the `plugin` project target Java 8 for its release output.

For example, you could have a shared [script plugin](https://docs.gradle.org/current/userguide/plugins.html#sec:script_plugins) for your build script(s) that looks like as follows:

```kts
import org.jetbrains.kotlin.gradle.dsl.JvmTarget
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

// Less verbose than `java.toolchain` but assumes that it's applied to a project
// with `org.jetbrains.kotlin.*` plugin applied, which you would, or the above
// `import` directives won't make sense.
kotlin.jvmToolchain(17)

internal object Build {
    val kotlinJvmTarget = JvmTarget.JVM_1_8
    const val javaReleaseOpt = 8
}
tasks.withType<KotlinCompile>().configureEach {
    compilerOptions.jvmTarget.set(Build.kotlinJvmTarget)
}
tasks.withType<JavaCompile>().configureEach {
    options.release.set(Build.javaReleaseOpt)
}
```

If the above is too much noise, or, if you want it to be shared across your other projects that require Java 8 as output, you could apply the above as a shared script plugin. Also, I think you could simplify the above by inferring `Build.kotlinJvmTarget` from `Build.javaReleaseOpt`, but at the moment, I couldn't, because I'm not currently using my IDE.

> If you need it applied as a shared script plugin, you could further enhance it by having the Java versions stored in your TOML or a subproject-scoped `gradle.properties` (accessed via `project.extra`). The latter option gives you flexibility in case you want to customize the Java target on the project you want to apply the shared script into. You could even combine both (i.e., access via `project.extra` if present, but fallback to TOML if otherwise). That way, you could have a single script for both your plugin and your Android demo project – by having each set up a project-scoped `gradle.properties` containing only the target Java version as its only property. An alternative to the project-scoped `gradle.properties` is to set up the target Java version via `project.extra["*"] = <java version>` prior to applying the shared script plugin.

Now, I used `JavaCompile.options.release` (instead of `JavaCompile.[source/target]Compatibility`) because it also conveniently sets up the `bootclasspath` (see, [javac - What is the --release flag in the Java 9 compiler? - Stack Overflow](https://stackoverflow.com/questions/43102787/what-is-the-release-flag-in-the-java-9-compiler)), given that with my suggestion, you'll be using a `javac` from a JDK higher than 9 (for your `plugin` project, even if the output is 8).

Also, you could indeed just set up the toolchain on the `plugin` project (and don't forget Gradle's `foojay` plugin in the settings script too). Then in that case, just remove the `kotlin.jvmToolchain(17)` line from the shared script above, and have it simply in your `plugin` project's build script exclusively.

---

Also, a word of caution when using `java.[source/target]Compatibility` (instead of `JavaCompile.[source/target]Compatibility`) with `java.toolchain`: the Gradle docs for [`java.toolchain`](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.JavaPluginExtension.html#org.gradle.api.plugins.JavaPluginExtension:toolchain) says the following:

> Configuring a toolchain cannot be used together with `sourceCompatibility` or `targetCompatibility` on this *[JavaPluginExtension]* extension. Both values will be sourced from the toolchain.

(Word in brackets added by me – for clarity.)

However, the above is probably just a typo from the docs, since in my experience, and also, by looking [at the source code](https://github.com/gradle/gradle/blob/v8.1.1/subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/DefaultJavaPluginExtension.java#L119), setting `sourceCompatibility` and/or `targetCompatibility` still works even with the toolchain configured (and without warnings being logged too). When I find the time, I might look for a relevant issue for that in the Gradle repo, and perhaps report it as a new issue to them if I couldn't find anything.

**Comment by gmazzo** [comment]
After I very looong time of "considering" this PR, the amount of changes to have a unified `foojay` setup on both build to setup the JDK on demand is just not worth it. It lost momentum after the recent changes where it got unified to JDK11 in the TOML file.

Thanks so much for the time you spent on this, and mostly for all the documentation you have provided.
